test/e2e/workload_clusters_test.go

```diff
@@ -36,6 +37,55 @@ func runWorkloadClusterFlowWithGitOps(test *framework.MulticlusterE2ETest, clust
 	test.DeleteManagementCluster()
 }
 
+func runWorkloadClusterUpgradeFlowCheckWorkloadRollingUpgrade(test *framework.MulticlusterE2ETest, clusterOpts ...framework.ClusterE2ETestOpt) {
+	latest := latestMinorRelease(test.T)
+	test.CreateManagementCluster(latest.Version, framework.ExecuteWithEksaRelease(latest))
+	test.RunInWorkloadClusters(func(w *framework.WorkloadCluster) {
+		w.GenerateClusterConfigForVersion(latest.Version)
+		w.CreateCluster(framework.ExecuteWithEksaRelease(latest))
+	})
+	preUpgradeMachines := make(map[string][]string, 0)
+	for key, workloadCluster := range test.WorkloadClusters {
+		test.T.Logf("Capturing CAPI machines for cluster %v", workloadCluster)
+		mdName := fmt.Sprintf("%s-%s", workloadCluster.ClusterName, "md-0")
+		test.ManagementCluster.WaitForMachineDeploymentReady(mdName)
+		preUpgradeMachines[key] = test.ManagementCluster.GetCapiMachinesForCluster(workloadCluster.ClusterName)
+	}
+	test.ManagementCluster.UpgradeCluster(clusterOpts)
+	test.T.Logf("Waiting for EKS-A controller to reconcile clusters")
+	time.Sleep(2 * time.Minute) // Time for new eks-a controller to kick in and potentially trigger rolling upgrade
+	for key, workloadCluster := range test.WorkloadClusters {
+		test.T.Logf("Capturing CAPI machines for cluster %v", workloadCluster)
+		postUpgradeMachines := test.ManagementCluster.GetCapiMachinesForCluster(workloadCluster.ClusterName)
+		if !areStringListsEquivalent(preUpgradeMachines[key], postUpgradeMachines) {
+			test.T.Fatalf("Found CAPI machines of workload cluster were changed after upgrading management cluster")
+		}
+	}
+	test.RunInWorkloadClusters(func(w *framework.WorkloadCluster) {
+		w.DeleteCluster()
+	})
+	test.DeleteManagementCluster()
+}
+
+func areStringListsEquivalent(list1 []string, list2 []string) bool {
```

Comment:
If we make `postUpgradeMachines` a hashmap based on the keys, then we can avoid this double nested for loop right? That's how I am understanding this. We can also maybe call this function `validateMachinesNotChanged` and then generalize it later if we need to use this functionality in other files.

Reply:
Will rename the function to be less generic and also converting the machines to be a map from name to machine